### ISSUE TITLE
Some tracer level is None when vmap'ing a custom_jvp

### ISSUE THREAD

**Issue by brianwa84** [google/jax#3822]
```python
!pip install -U -q jax jaxlib
!pip freeze

import jax
import jax.numpy as np

alpha = np.float32(2.)

def sample(seed):
  @jax.custom_jvp
  def f(alpha):
    return jax.random.gamma(seed, alpha, shape=[])

  @f.defjvp
  def f_jvp(primal, tangent):
    alpha = primal
    dalpha = tangent
    sample = f(alpha)
    partial_alpha = jax.lax.random_gamma_grad(alpha, sample)
    return sample, partial_alpha * dalpha
  return f(alpha)

jax.vmap(sample)(jax.random.split(jax.random.PRNGKey(1), 3))
```
=>
```none
...
jax==0.1.72
jaxlib==0.1.51
...
/usr/local/lib/python3.6/dist-packages/jax/core.py in <listcomp>(.0)
   1078   todo = []
   1079   while True:
-> 1080     tracers = [x for x in outs if isinstance(x, Tracer) and x._trace.level > level]
   1081     if tracers:
   1082       ans = max(tracers, key=lambda x: x._trace.level)

TypeError: '>' not supported between instances of 'int' and 'NoneType'
```

**Comment by mattjj** [comment]
I think the issue is that `custom_jvp` functions and their corresponding JVP rules need to be defined at the top level rather than inside function bodies, i.e. they can't be indented. (It is possible this is requirement not explained in the documentation anywhere! I need to double check that it's actually true, since it might not be, then document it.)

That's why I changed the issue labels anyway. But then I realized that this particular error is unexpected, so I need to investigate more...

**Comment by mattjj** [comment]
After #3830, the error message for this code is:

> jax.core.UnexpectedTracerError: Encountered an unexpected tracer in custom_jvp function f. custom_jvp functions and their corresponding differentiation rules must be defined at the top-level, i.e. not within the body of another function to be transformed.

An impovement! Still some todo items for me though.

**Comment by mattjj** [comment]
I'd like to look again at whether it's possible to support closed-over tracers in custom_jvp/vjp functions, like we do for `jit` and `lax.scan` and the rest...

**Comment by mattjj** [comment]
This was fixed by #4008 (which got merged, though our new source management system didn't mark the PR as closed).
